### PR TITLE
[ASCII-2052] Fix races in expvar http handler when returning a slice and marshalling

### PR DESCRIPTION
**Pull request by pgimalac**
### What does this PR do?

Clone slices returned in `expvar.Func` to avoid races when marshalling.

### Motivation

We've seen a panic in the expvar http handler where a slice was updated while the expvar was being marshalled, this PR clones all such cases to avoid another occurence.

### Additional Notes

Here is an abstract of the stack trace:
```
panic
	/usr/local/go/src/runtime/panic.go:770
reflect.Value.Index
	/usr/local/go/src/reflect/value.go:1430
encoding/json.arrayEncoder.encode
	/usr/local/go/src/encoding/json/encode.go:847
encoding/json.sliceEncoder.encode
	/usr/local/go/src/encoding/json/encode.go:820
encoding/json.structEncoder.encode
	/usr/local/go/src/encoding/json/encode.go:704
encoding/json.ptrEncoder.encode
	/usr/local/go/src/encoding/json/encode.go:876
encoding/json.mapEncoder.encode
	/usr/local/go/src/encoding/json/encode.go:761
encoding/json.mapEncoder.encode
	/usr/local/go/src/encoding/json/encode.go:761
encoding/json.(*encodeState).reflectValue
	/usr/local/go/src/encoding/json/encode.go:321
encoding/json.(*encodeState).marshal
	/usr/local/go/src/encoding/json/encode.go:297
encoding/json.Marshal
	/usr/local/go/src/encoding/json/encode.go:163
expvar.Func.String
	/usr/local/go/src/expvar/expvar.go:302
expvar.(*Map).appendJSONMayExpand.func3
	/usr/local/go/src/expvar/expvar.go:156
expvar.(*Map).Do
	/usr/local/go/src/expvar/expvar.go:265
```

We can see that it happens while marshalling the return value of some `expvar.Func`, and looking at the json encoding stack trace the actual type seem to be a map of map to pointer to struct containing a slice (ie. `map[A]map[B]*struct{[]C}`).

This type matches the func added in [pkg/collector/runners/expvars/expvars.go](https://github.com/DataDog/datadog-agent/blob/065aa613df6f56d813397283b9273f70f8975384/pkg/collector/runner/expvars/expvars.go#L49), which returns a `map[string]map[checkid.ID]*checkstats.Stats`, where `checkstats.Stats` contains a slice.

### CODE REVIEW

File: pkg/collector/check/stats/stats.go [hunk 1]
```diff
@@ -258,6 +260,41 @@ func (cs *Stats) SetStateCancelling() {
...
+		AverageExecutionTime:     cs.AverageExecutionTime,
+		LastExecutionTime:        cs.LastExecutionTime,
+		LastSuccessDate:          cs.LastSuccessDate,
+		LastError:                cs.LastError,
+		LastDelay:                cs.LastDelay,
+		LastWarnings:             slices.Clone(cs.LastWarnings),
+		UpdateTimestamp:          cs.UpdateTimestamp,
+		telemetry:                cs.telemetry,
+	}
+	copy(cs.ExecutionTimes[:], cloned.ExecutionTimes[:])
```
Comment (vickenty) [inline pkg/collector/check/stats/stats.go:294]: This arguments should be the other way around: https://pkg.go.dev/builtin#copy

Would it make sense to use deepcopy package (already a dependency) to handle this and other deep-copying cases as well?

Comment (pgimalac) [reply pkg/collector/check/stats/stats.go:294]: Oh nice catch !

> Would it make sense to use deepcopy package to handle this

Sure, that would be safer !

**Review by GustavoCaso — COMMENTED** [review]
Not a blocker, but more of a train of thought 🚋 
Would be interesting to move all expvar logic to a core component, that way each sub component could create the expvars using a the one component, and for reading we would be able to control the concurrency access, that way we do not have to patch here and there.

**Review by GustavoCaso — COMMENTED** [review]
@pgimalac would it be possible to add a regression test to ensure concurrent access to those expvar values do not cause a panic?

**Comment by pgimalac** [comment]
@GustavoCaso I added a test to trigger the race which caused a panic (before the fix).
I can't really test every `expvar.Func` however, they all behave differently and they would need a custom test for each...

**Comment by AliDatadog** [comment]
> Would be interesting to move all expvar logic to a core component

IMO we should fully get rid of expvar

**Comment by GustavoCaso** [comment]
> IMO we should fully get rid of expvar

I'm onboard of that train 100%.

Unfortunately, I'm not sure if would be that easy, as they are accesible via an HTTP endpoint. Some customers, might have scripts that consume them and create reports or any other automation workflow.

File: pkg/collector/runner/expvars/expvars_test.go [hunk 2]
```diff
@@ -398,3 +400,41 @@ func TestExpvarsToplevelKeys(t *testing.T) {
 		changeAndAssertExpvarValue(t, keyName, setter, getters[keyName], 3, 4)
 	}
 }
+
+func TestGetCheckStatsRace(t *testing.T) {
+	Reset()
```
Comment (GustavoCaso) [inline pkg/collector/runner/expvars/expvars_test.go:405]: Should we use the test help function `setUp()`?

Comment (pgimalac) [reply pkg/collector/runner/expvars/expvars_test.go:405]: I didn't really understand the point of that function, it's purely just `Reset()`, I thought it was clearer to call it explicitly
Not all tests even call it by the way 🙃

Comment (GustavoCaso) [reply pkg/collector/runner/expvars/expvars_test.go:405]: > I thought it was clearer to call it explicitly

It does makes sense. We could improve the rest of the test by adding the same cleanup code you added nd remove the `setup()` function all together 

```go
Reset()
t.Cleanup(Reset)
```

That way we standardize the test 😄

**Review by vickenty — APPROVED** [review]

**Review by GustavoCaso — APPROVED** [review]

**Review by adel121 — COMMENTED** [review]
LGTM for container platform owned files.

**Review by adel121 — APPROVED** [review]

**Review by daniel-taf — APPROVED** [review]

**Review by knusbaum — APPROVED** [review]

**Pull request merged**